src/wp-includes/interactivity-api/class-wp-interactivity-api.php

```diff
@@ -0,0 +1,959 @@
+	 * @return string The new style attribute value after the specified property has been added, updated or removed.
+	 */
+	private function merge_style_property( string $style_attribute_value, string $style_property_name, $style_property_value ): string {
+		$style_assignments    = explode( ';', $style_attribute_value );
+		$result               = array();
+		$style_property_value = ! empty( $style_property_value ) ? rtrim( trim( $style_property_value ), ';' ) : null;
+		$new_style_property   = $style_property_value ? $style_property_name . ':' . $style_property_value . ';' : '';
+
+		// Generates an array with all the properties but the modified one.
+		foreach ( $style_assignments as $style_assignment ) {
+			if ( empty( trim( $style_assignment ) ) ) {
+				continue;
+			}
+			list( $name, $value ) = explode( ':', $style_assignment );
+			if ( trim( $name ) !== $style_property_name ) {
+				$result[] = trim( $name ) . ':' . trim( $value ) . ';';
+			}
+		}
+
+		// Adds the new/modified property at the end of the list.
+		$result[] = $new_style_property;
+
+		return implode( '', $result );
+	}
+
+	/**
+	 * Processes the `data-wp-text` directive.
+	 *
+	 * It updates the inner content of the current HTML element based on the
+	 * evaluation of its associated reference.
+	 *
+	 * @since 6.5.0
+	 *
+	 * @param WP_Interactivity_API_Directives_Processor $p               The directives processor instance.
+	 * @param array                                     $context_stack   The reference to the context stack.
+	 * @param array                                     $namespace_stack The reference to the store namespace stack.
+	 */
+	private function data_wp_text_processor( WP_Interactivity_API_Directives_Processor $p, array &$context_stack, array &$namespace_stack ) {
+		if ( ! $p->is_tag_closer() ) {
+			$attribute_value = $p->get_attribute( 'data-wp-text' );
+			$result          = $this->evaluate( $attribute_value, end( $namespace_stack ), end( $context_stack ) );
+
+			/*
+			 * Follows the same logic as Preact in the client and only changes the
+			 * content if the value is a string or a number. Otherwise, it removes the
+			 * content.
+			 */
+			if ( is_string( $result ) || is_numeric( $result ) ) {
+				$p->set_content_between_balanced_tags( esc_html( $result ) );
+			} else {
+				$p->set_content_between_balanced_tags( '' );
+			}
+		}
+	}
+
+	/**
+	 * Returns the CSS styles for animating the top loading bar in the router.
+	 *
+	 * @since 6.5.0
+	 *
+	 * @return string The CSS styles for the router's top loading bar animation.
+	 */
+	private function router_animation_styles(): string {
+		return <<<CSS
+			.wp-interactivity-router-loading-bar {
+				position: fixed;
+				top: 0;
+				left: 0;
+				margin: 0;
+				padding: 0;
+				width: 100vw;
+				max-width: 100vw !important;
+				height: 4px;
+				background-color: var(--wp--preset--color--primary, #000);
```

Comment:
How is it ensured that this custom property is always defined?

The `wp-interactivity-router-animations` stylesheet would need to have a corresponding dependency, no?

Reply:
Not really. If it's a block theme, it'll get its primary color. If it's a classic theme, it'll initially show as black (`#000`), but the theme can change it by defining the custom property.

Reply:
`--wp--preset--color--primary` is defined in Global Styles:

<img width="1207" alt="Screenshot 2024-02-07 at 10 48 38" src="https://github.com/WordPress/wordpress-develop/assets/699132/b5ccdabd-4931-4a79-b419-c3340308431a">

@oandregal, to use `--wp--preset--color--primary`, does the CSS (not connected to blocks) need to define the dependency on `global-styles`?

Reply:
AFAIK, [core](https://github.com/WordPress/wordpress-develop/blob/master/src/wp-includes/theme.json#L116) doesn't define any "primary" and "primary" is not an abstraction themes have to provide either (see [TT4](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-content/themes/twentytwentyfour/theme.json#L103), the latest default theme). Additionally, the CSS Custom Properties generated by `theme.json` aren't meant to be public API.

I'm not sure what this is for so I can't provide alternatives, but TLDR, we shouldn't rely on `--wp--preset-color--*`.

Reply:
cc @aaronrobertshaw as his knowledge on this area is more up to date than mine is.

Reply:
> I'm not sure what this is for so I can't provide alternatives

This is the color of the top loading bar that shows on client-side navigation.

By default it's black, but if the theme has defined a primary color, it uses that instead:

https://github.com/WordPress/wordpress-develop/assets/3305402/831bc65b-a3e6-434a-8b1d-21847b210a27

Reply:
> This is the color of the top loading bar that shows on client-side navigation.

The native way for themes to address that would be converting it to a block or an element. For example, via `theme.json` a theme would do something along these lines:

```json
{
  "styles": {
    "elements": {
      "navigation-loading-bar": {
        "color": "VALUE"
      }
    }
  }
}
```

or 

```json
{
  "styles": {
    "blocks": {
      "core/navigation-loading-bar": {
        "color": "VALUE"
      }
    }
  }
}
```

Reply:
It's not a block or an element per se as this type of client navigation can be triggered by multiple blocks present on the page or even classic themes, but I like the idea of incorporating it into the `theme.json` mechanism.

We can keep the black for WP 6.5, remove the primary custom property, and research how to add it to the `theme.json` for WP 6.6. Themes can always override it with CSS, so they are not limited to black.

Reply:
Fixed in https://github.com/WordPress/wordpress-develop/pull/5953/commits/76dea4f71d9ede63f6b2a2a71f75958222fa4bbc.